src/tools/wasm-split/instrumenter.cpp

```diff
@@ -142,18 +171,76 @@ void Instrumenter::addProfileExport() {
   // Write the hash followed by all the time stamps
   Expression* writeData =
     builder.makeStore(8, 0, 1, getAddr(), hashConst(), Type::i64);
-
   uint32_t offset = 8;
-  for (const auto& global : functionGlobals) {
-    writeData = builder.blockify(
-      writeData,
-      builder.makeStore(4,
-                        offset,
-                        1,
-                        getAddr(),
-                        builder.makeGlobalGet(global, Type::i32),
-                        Type::i32));
-    offset += 4;
+
+  switch (options.storageKind) {
+    case WasmSplitOptions::StorageKind::InGlobals: {
+      for (const auto& global : functionGlobals) {
+        writeData = builder.blockify(
+          writeData,
+          builder.makeStore(4,
+                            offset,
+                            1,
+                            getAddr(),
+                            builder.makeGlobalGet(global, Type::i32),
+                            Type::i32));
+        offset += 4;
+      }
+      break;
+    }
+    case WasmSplitOptions::StorageKind::InMemory: {
+      Index funcIdxVar =
+        Builder::addVar(writeProfile.get(), "funcIdx", Type::i32);
+      auto getFuncIdx = [&]() {
+        return builder.makeLocalGet(funcIdxVar, Type::i32);
+      };
+      // (block $outer
+      //   (loop $l
+      //     (br_if $outer (i32.eq (local.get $fucIdx) (i32.const numFuncs))
+      //     (i32.store offset=8
+      //       (i32.add
+      //         (local.get $addr)
+      //         (i32.mul (local.get $funcIdx) (i32.const 4))
+      //       )
+      //       (i32.atomic.load8_u (local.get $funcIdx))
+      //     )
+      //     (local.set $funcIdx
+      //      (i32.add (local.get $fundIdx) (i32.const 1)
+      //     )
+      //     (br $l)
+      //   )
+      // )
+      writeData = builder.blockify(
+        writeData,
+        builder.makeBlock(
+          "outer",
+          builder.makeLoop(
+            "l",
+            builder.blockify(
+              builder.makeBreak(
+                "outer",
+                nullptr,
+                builder.makeBinary(EqInt32,
+                                   getFuncIdx(),
+                                   builder.makeConst(uint32_t(numFuncs)))),
+              builder.makeStore(
+                4,
+                offset,
+                4,
+                builder.makeBinary(
+                  AddInt32,
+                  getAddr(),
+                  builder.makeBinary(
+                    MulInt32, getFuncIdx(), builder.makeConst(uint32_t(4)))),
+                builder.makeAtomicLoad(1, 0, getFuncIdx(), Type::i32),
+                Type::i32),
+              builder.makeLocalSet(
+                funcIdxVar,
+                builder.makeBinary(
+                  AddInt32, getFuncIdx(), builder.makeConst(uint32_t(1)))),
+              builder.makeBreak("l")))));
+      break;
+    }
```

Comment:
Is `writeData` needed for InMemory - I think it can just be read from where it is? Unless I guess you want a uniform interface.

Reply:
Yes, I definitely want to give a uniform interface. Also, the in-memory representation (1 byte per function) is not the same as the profile representation (hash + 4 bytes per function), and I'd also rather keep a uniform profile format for now as well.